### PR TITLE
Partest fails if nothing to compile

### LINKED ISSUE #12769: partest of empty dir
## Problem

An empty `pos` dir passes, an empty `run` dir crashes where
```
last test / IntegrationTest / testOnly
```
shows
```
[error] Uncaught exception when running partest: java.util.concurrent.ExecutionException: java.lang.RuntimeException: Error running .../scala/test/files/run/ZZZZ
[scroll way down]
[error] Caused by: sbt.ForkMain$ForkError: java.nio.file.NoSuchFileException: .../scala/test/files/run/ZZZZ-run.log
```

I was going to make `ZZZZ.scala` into `ZZZZ/*.scala`, changed my mind. It takes the empty dir and ignores the `ZZZZ.scala`, but really if it decides it has nothing to compile, it should fail the test instead of either passing or crashing.

Someday, when Scala 3 seeks to re-integrate partest, they'll say, "It even fails nicely on an empty test directory."

### PR DESCRIPTION
**Pull request by som-snytt**
Also, when crashing, create & append to log file (which may not exist because crashing).

Fixes scala/bug#12769

### CODE REVIEW

**Comment by som-snytt** [comment]
Partest doesn't have meta-tests, although you could rig `partest` tests to only check checkfiles, that sounds easy.

In lieu of that, did you double-check the behaviors manually?

```
# starting 1 test in pos
!! 1 - pos/empty                                 [nothing to compile]
# 0/1 passed, 1 failed in pos

# starting 1 test in neg
!! 1 - neg/empty                                 [nothing to compile]
# 0/1 passed, 1 failed in neg

# starting 1 test in run
Note: test execution will be non-parallel under -Dpartest.exec.in.process
!! 1 - run/empty                                 [nothing to compile]
# 0/1 passed, 1 failed in run
```
and failing tests fail, and flailing tests as well
```
# starting 1 test in pos
!! 1 - pos/badpos.scala                          [compilation failed with 1 errors]
##### Log file '.../scala/test/files/pos/badpos-pos.log' from failed test #####

badpos.scala:2: error: expected class or object definition
clas C
^
1 error
# 0/1 passed, 1 failed in pos

# starting 1 test in neg
!! 1 - neg/badneg.scala                          [expected compilation failure]
##### Log file '.../scala/test/files/neg/badneg-neg.log' from failed test #####

# 0/1 passed, 1 failed in neg
# starting 1 test in run
Note: test execution will be non-parallel under -Dpartest.exec.in.process
!! 1 - run/badrun.scala                          [output differs]
% diff .../scala/test/files/run/badrun.check .../scala/test/files/run/badrun-run.log--- empty
+++ badrun-run.log
@@ -1,0 +1,49 @@
+scala.NotImplementedError: an implementation is missing

```

File: src/partest/scala/tools/partest/nest/Runner.scala [hunk 1]
```diff
@@ -552,25 +545,30 @@ class Runner(val testInfo: TestInfo, val suiteRunner: AbstractRunner) {
...
 
-    compilationRounds(testFile).find(r => !r.result.isOk || r.result.isSkipped).map(checked).getOrElse(genFail("expected compilation failure"))
+    runTestCommon(checked, expectCompile = false)()
   }
 
   // run compilation until failure, evaluate `andAlso` on success
-  def runTestCommon(andAlso: => TestState = genPass()): TestState = runInContext {
-    // DirectCompiler already says compilation failed
-    val res = compilationRounds(testFile).find(r => !r.result.isOk || r.result.isSkipped).map(_.result).getOrElse(genPass())
-    res andAlso andAlso
```
Comment (som-snytt) [inline src/partest/scala/tools/partest/nest/Runner.scala:573]: one must sacrifice some poetry. Tomorrow, and tomorrow, and tomorrow, creeps in this petty pace from day to day...

**Comment by som-snytt** [comment]
Somewhere I have a branch that adds metatests under `test/files/partest/pos` etc. My timebox expired. (Presumably it didn't work immediately.)

Vulpix also will report success if no tests were run in any of the "groups". I find that quite annoying. For me, it's hard enough to figure out, "How do I run this particular test?"

The killer feature would be, `test --explain path/mytest.scala` and the test rig responsible would spit out instructions.

**Review by lrytz — APPROVED** [review]

**Comment by SethTisue** [comment]
merging since "internal"

**Comment by som-snytt** [comment]
Moved that metatest facility and another commit to a new branch for later.
```
commit 4d0600e3d59cf5f9e339ceefb73dcb2d9424c859 (HEAD -> tweak/partest-rounds)
Author: Som Snytt <som.snytt@gmail.com>
Date:   Mon May 1 23:35:18 2023 -0700

    Remove obsolete OnlyScala 3rd round

commit 63628d2a4dfc7b9d137bc4b06f3a0c3e255378fb
Author: Som Snytt <som.snytt@gmail.com>
Date:   Fri Apr 7 12:27:12 2023 -0700

    Partest tests partest
```

**Pull request merged**